### PR TITLE
Refactor request validation logic to middleware components

### PR DESCRIPTION
**Pull request by mtrang1263**
### Needed for [BCDA-4240](https://jira.cms.gov/browse/BCDA-4240)

We want to add additional handler logic to serve out ALR data. As it stands, we would have to duplicate quite a bit of code to achieve the same requests logic (rate limiting, request validation, etc.) in order to achieve this new handler logic. With this PR, we move out this logic into new middleware components.

By using the middleware components, we can ensure all of our export jobs have the consistent validation behavior.

### Change Details
1. Move request validation logic to new middleware handler (ValidateRequestURL)
2. Move 429 logic to new middleware handler (CheckConcurrentJobs)
3. Clean up tests in v1/api_test.go and requests.go based on the logic changes.
 

### Security Implications

- [ ] new software dependencies
- [ ] security controls or supporting software altered
- [ ] new data stored or transmitted
- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Acceptance Validation
1. CI Passes.
2. Successfully deployed and ran smoke tests on dev environment.

### CODE REVIEW

File: bcda/api/requests_test.go [hunk 1]
```diff
@@ -123,159 +127,80 @@ func (s *RequestsTestSuite) TestRunoutDisabled() {
 	s.Contains(string(body), "Invalid group ID")
 }
 
-func (s *RequestsTestSuite) TestInvalidRequests() {
```
Comment (mtrang1263) [inline bcda/api/requests_test.go:126]: Replaced by test in [validation_test.go](https://github.com/CMSgov/bcda-app/pull/673/files#diff-a03a477e3e4270213962d11eb6f2f8c5386f25cac2ac4976a001dbf8d0fbd45cR43)

File: bcda/api/requests_test.go [hunk 2]
```diff
@@ -123,159 +127,80 @@ func (s *RequestsTestSuite) TestRunoutDisabled() {
...
-		s.T().Run(fmt.Sprintf("%s-patient", tt.name), func(t *testing.T) {
-			rr := httptest.NewRecorder()
-			h.BulkGroupRequest(rr, req)
-			assert.Equal(t, http.StatusBadRequest, rr.Code)
-			assert.Contains(t, rr.Body.String(), tt.errMsg)
-		})
 	}
-}
 
-func (s *RequestsTestSuite) TestCheck429() {
```
Comment (mtrang1263) [inline bcda/api/requests_test.go:206]: Replace by test in [ratelimit_test.go](https://github.com/CMSgov/bcda-app/pull/673/files#diff-75b0d1f2ed7de3581ea31fc1ebe5a2747d66a4e5f65fe8034494cc040b131d39R46)

File: bcda/api/requests_test.go [hunk 3]
```diff
@@ -123,159 +127,80 @@ func (s *RequestsTestSuite) TestRunoutDisabled() {
...
+			rr := httptest.NewRecorder()
+			h.BulkPatientRequest(rr, s.genPatientRequest(rp))
+			assert.Equal(s.T(), http.StatusAccepted, rr.Code)
+		}
 	}
 }
 
-// TestBulkRequestWithOldJobPaths verifies that bulk requests for a caller with older jobs
-// whose paths are no longer valid will still be accepted.
-func (s *RequestsTestSuite) TestBulkRequestWithOldJobPaths() {
```
Comment (mtrang1263) [inline bcda/api/requests_test.go:240]: Replace by test in [ratelimit_test.go](https://github.com/CMSgov/bcda-app/pull/673/files#diff-75b0d1f2ed7de3581ea31fc1ebe5a2747d66a4e5f65fe8034494cc040b131d39R46)

File: bcda/web/middleware/middleware.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package middleware
```
Comment (mtrang1263) [inline bcda/web/middleware/middleware.go:1]: Replaced bcda/web/middleware.go

File: bcda/web/middleware/middleware_test.go [hunk 5]
```diff
@@ -31,80 +31,6 @@ func (s *MiddlewareTestSuite) SetupTest() {
...
-
-	resp, err := client.Do(req)
-	if err != nil {
-		log.Fatal(err)
-	}
-
-	assert.Equal(s.T(), 200, resp.StatusCode)
-}
-
-func (s *MiddlewareTestSuite) TestValidateBulkRequestHeadersInvalidAccept() {
```
Comment (mtrang1263) [inline bcda/web/middleware/middleware_test.go:53]: Test moved to validation_test.go

File: bcda/web/middleware/middleware_test.go [hunk 6]
```diff
@@ -31,80 +31,6 @@ func (s *MiddlewareTestSuite) SetupTest() {
...
-
-	resp, err = client.Do(req)
-	if err != nil {
-		log.Fatal(err)
-	}
-
-	assert.Equal(s.T(), 400, resp.StatusCode)
-}
-
-func (s *MiddlewareTestSuite) TestValidateBulkRequestHeadersInvalidPrefer() {
```
Comment (mtrang1263) [inline bcda/web/middleware/middleware_test.go:81]: Test moved to validation_test.go

File: bcda/api/requests.go [hunk 7]
```diff
@@ -155,39 +156,6 @@ func (h *Handler) bulkRequest(resourceTypes []string, w http.ResponseWriter, r *
 
 	acoID := uuid.Parse(ad.ACOID)
 
-	// If we really do find this record with the below matching criteria then this particular ACO has already made
-	// a bulk data request and it has yet to finish. Users will be presented with a 429 Too-Many-Requests error until either
-	// their job finishes or time expires (+24 hours default) for any remaining jobs left in a pending or in-progress state.
-	// Overall, this will prevent a queue of concurrent calls from slowing up our system.
-	// NOTE: this logic is relevant to PROD only; simultaneous requests in our lower environments is acceptable (i.e., shared opensbx creds)
-	if conf.GetEnv("DEPLOYMENT_TARGET") == "prod" {
```
Comment (mtrang1263) [inline bcda/api/requests.go:163]: Replace by [ratelimit.go](https://github.com/CMSgov/bcda-app/pull/673/files#diff-a52b8517ff096bf2da0c14f70494c8bca8dbfc20e6d1f01615d0858ce4ea6b28R33)

File: bcda/api/requests.go [hunk 8]
```diff
@@ -334,130 +290,15 @@ func (h *Handler) bulkRequest(resourceTypes []string, w http.ResponseWriter, r *
 	}
 }
 
-func (h *Handler) validateRequest(r *http.Request) ([]string, *fhirmodels.OperationOutcome) {
```
Comment (mtrang1263) [inline bcda/api/requests.go:337]: Replaced by validation.go

**Review by mtrang1263 — COMMENTED** [review]
All tests that were removed in v1/api_test.go were replaced by [requests_test.go#TestRequests](https://github.com/CMSgov/bcda-app/pull/673/files#diff-7fc3cc46a29844127337e64bc5dbd5d50fbfa043073ae134ea665a8c07cedd9bR131)

File: bcda/web/middleware/validation.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+type RequestParameters struct {
+	Since         time.Time
+	ResourceTypes []string
+	Version       string // e.g. v1, v2
+}
+
+// requestkey is an unexported context key to avoid collisions
+type requestkey int
+
+var rk requestkey
```
Comment (CosmicSyntax) [inline bcda/web/middleware/validation.go:29]: Make this a const? Just in case

File: bcda/api/requests.go [hunk 10]
```diff
@@ -112,38 +109,42 @@ func (h *Handler) BulkGroupRequest(w http.ResponseWriter, r *http.Request) {
...
-		responseutils.WriteError(oo, w, http.StatusBadRequest)
-	}
-
 	if ad, err = readAuthData(r); err != nil {
 		oo := responseutils.CreateOpOutcome(fhircodes.IssueSeverityCode_ERROR, fhircodes.IssueTypeCode_EXCEPTION, responseutils.TokenErr, "")
 		responseutils.WriteError(oo, w, http.StatusUnauthorized)
 		return
 	}
 
+	rp, ok := middleware.RequestParametersFromContext(r.Context())
```
Comment (CosmicSyntax) [inline bcda/api/requests.go:130]: Change to ctx to be consistent?

File: bcda/web/middleware/validation.go (hunk 9 above)
Comment (mtrang1263) [reply bcda/web/middleware/validation.go:29]: great call out! I misread Go's guidance here and was using a var. Ref: https://blog.golang.org/context

**Review by CosmicSyntax — APPROVED** [review]
This looks awesome! Amazing work!

LGTM!

**Pull request merged**